system-addon/lib/PlacesFeed.jsm

```diff
@@ -187,7 +193,40 @@ class PlacesFeed {
     Services.obs.addObserver(this, LINK_BLOCKED_EVENT);
   }
 
+  /**
+   * setTimeout - A custom function that creates an nsITimer used to debounce bookmark notifications
+   *
+   * @param {func} callback       A function to be executed after the timer expires
+   * @param {int}  delay          The time (in ms) the timer should wait before the function is executed
+   */
+  setTimeout(callback, delay) {
+    let timer = Cc["@mozilla.org/timer;1"].createInstance(Ci.nsITimer);
+    timer.initWithCallback(callback, delay, Ci.nsITimer.TYPE_ONE_SHOT);
+    return timer;
+  }
+
+  customDispatch(action) {
+    // If we are changing many links at once, debounce this action and only dispatch
+    // at the end
+    if (action.type === at.PLACES_LINKS_CHANGED) {
+      if (this.placesChangedTimer) {
+        this.placesChangedTimer.delay = PLACES_LINKS_CHANGED_DEBOUNCE_TIME;
+      } else {
+        this.placesChangedTimer = this.setTimeout(() => {
+          this.placesChangedTimer = null;
+          this.store.dispatch(ac.OnlyToMain(action));
```

Comment:
Oh only to main is kinda tricky, but that's fine. Hopefully callers don't accidentally use some other action creator…